src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java

```diff
@@ -125,68 +114,64 @@ public MyCollectionsApiController(ObjectMapper objectMapper, HttpServletRequest
-  	
     	if (!lidvid.contains("::")) lidvid = this.productBO.getLatestLidVidFromLid(lidvid);
     	MyCollectionsApiController.log.info("request collection lidvid, collections children: " + lidvid);
-         
-    	try {
-	    	Products products = new Products();
-	    	
-	    	HashSet<String> uniqueProperties = new HashSet<String>();
-	    	
-	      	Summary summary = new Summary();
-	    	
-	    	summary.setStart(start);
-	    	summary.setLimit(limit);
-	    	
-	    	if (sort == null) {
-	    		sort = Arrays.asList();
-	    	}	
-	    	summary.setSort(sort);
-	    	
-	    	products.setSummary(summary);
-	    	
-	    	CollectionProductRefBusinessObject  collectionProductRefBO = new CollectionProductRefBusinessObject(this.esRegistryConnection);
-	    	CollectionProductRelationships collectionProductRelationships = collectionProductRefBO.getCollectionProductsIterable(lidvid, start, limit);
-	    	
-	    	for (EntityProduct eProd : collectionProductRelationships) {
-				  if (eProd != null) {
-		        	MyCollectionsApiController.log.info("request lidvdid: " + eProd.getLidVid() );
-		        	
-		    		Product product = ElasticSearchUtil.ESentityProductToAPIProduct(eProd, this.getBaseURL());
-				    		
-		    		Map<String, XMLMashallableProperyValue> filteredMapJsonProperties = ProductBusinessObject.getFilteredProperties(
-		    				eProd.getProperties(), 
-		    				fields,
-		    				null);
-		
-		    		uniqueProperties.addAll(filteredMapJsonProperties.keySet());
-		
-		    		if (!onlySummary) {
-		        		product.setProperties((Map<String, PropertyArrayValues>)(Map<String, ?>)filteredMapJsonProperties);
-		        		
-		        		products.addDataItem(product);
-		    		}
+
+    	int iteration=0,wsize=0;
+    	HashSet<String> uniqueProperties = new HashSet<String>();
+    	List<String> plidvids = new ArrayList<String>();
+    	List<String> wlidvids = new ArrayList<String>();
+    	Products products = new Products();
+      	Summary summary = new Summary();
+
+    	if (sort == null) { sort = Arrays.asList(); }	
+
+    	summary.setStart(start);
+    	summary.setLimit(limit);
+      	summary.setSort(sort);	
+    	products.setSummary(summary);
+
+    	for (final Map<String,Object> kvp : new ElasticSearchHitIterator(10, this.esRegistryConnection.getRestHighLevelClient(),
+				ElasticSearchRegistrySearchRequestBuilder.getQueryFieldFromKVP("collection_lidvid", lidvid, "product_lidvid",
+						this.esRegistryConnection.getRegistryRefIndex())))
+		{
+    		wlidvids.clear();
+    		wsize = 0;
+
+    		if (kvp.get("product_lidvid") instanceof String)
+			{ wlidvids.add(this.productBO.getLatestLidVidFromLid(kvp.get("product_lidvid").toString())); }
+			else
+			{
+				@SuppressWarnings("unchecked")
+				List<String> clids = (List<String>)kvp.get("product_lidvid");
+
+				if (start <= iteration || start < iteration+clids.size())
```

Comment:
This section of code is difficult to understand with multiple cases on an iteration counter. Could you add comments for each case ?

Is this logic re-used in the product of bundle method ? Since product of a bundle are the product of the collections of the bundles that would make sense to me.

Reply:
There is some overlap between bundle and collection, but it is not the same strictly speaking. The bundle version looks up the product_lidvid for may collections all at once while this one is a singular.